### PR TITLE
Add custom order display option

### PR DESCRIPTION
**Pull request by weeman1337**
* Fix contribute link
* Add the new *custom order* option
* Add support for the new sort setting
* Tweak the bookmark controller to support sort by `index`

I really do not see any chance me rewriting the app to the bookmarks API v2. So I extended the bookmark controller and added support for the new sort option.

Close issue #778

### CODE REVIEW

**Comment by weeman1337** [comment]
@marcelklehr could you send me info on about how to set up the translation for the new string? I could also provide the first translation.

**Comment by marcelklehr** [comment]
New strings are automatically picked up by the transifex bot and translatable via transifex.

**Comment by marcelklehr** [comment]
Sorry for the late reply. I'm not sure if I like this approach, as the server actually also stores folder's positions and the bookmarks endpoint is also used for other queries where sorting by index doesn't make sense.

> I really do not see any chance me rewriting the app to the bookmarks API v2.

The bookmarks UI already uses the v2 API, but it's using the private routes not the public routes. I probably should have explained that.

**Comment by weeman1337** [comment]
> I'm not sure if I like this approach

@marcelklehr how would you like it to be?

**Comment by marcelklehr** [comment]
I was thinking of using this endpoint: https://nextcloud-bookmarks.readthedocs.io/en/latest/folder.html#get-folder-s-contents

But that likely requires some more refactoring.

**Comment by weeman1337** [comment]
> I was thinking of using this endpoint: https://nextcloud-bookmarks.readthedocs.io/en/latest/folder.html#get-folder-s-contents

This endpoint doesn't support sorting and the returned object do not contain all the fields for doing it in the front end (index, number of clicks, creation time).

**Comment by weeman1337** [comment]
ping @marcelklehr

**Comment by marcelklehr** [comment]
> ping

Sorry :see_no_evil: 

> This endpoint doesn't support sorting and the returned objects do not contain all the fields for sorting the front end (index, number of clicks, creation time).

This endpoint already returns the ordered data, but is indeed missing some items.

> The order could be retrieved by GET /public/rest/v2/folder/(int: folder_id)/childorder but then still number of clicks and creation time is missing.

This endpoint is supposed to only provide the order while the actual data comes from the /folder and /bookmark endpoints.

I think a good way to implement "custom ordering" would be something like this

1) [if ordering != 'custom'] display bookmarks and folders as returned from the API with `sortby` param
2) [if ordering == 'custom'] fetch childorder endpoint for currently active folder and arrange already loaded bookmarks and folders according to that order.

What do you think?

**Comment by weeman1337** [comment]
Thanks for the answer.

> What do you think?

It is kind of inconsistent, but works. For an API client it would be much easier to pass the desired order as a query param to a single API endpoint.

> This endpoint already returns the ordered data, but is indeed missing some items.

It returns the "custom order", yes. But it is not possible to request a different one.

**Comment by marcelklehr** [comment]
Yeah the reason for this inconsistent design, is that the "custom order" is kind of part of the data model to be consistent with browser implementations, while all other possible orderings are just "ORDER BY" appendices on the bookmarks table. This is because the custom order explicitly "includes" folders, so you can e.g. sort 3 bookmarks to the top, then have two folders and then the rest of the bookmarks, which is not possible with the normal sort options (those are bookmarks-only).

**Comment by marcelklehr** [comment]
I've finally had some time to mull this over in my head and I think your approach is viable together with a small change in the way the results are displayed that I've committed to your branch just now. Let me know what you think :)

**Review by marcelklehr — APPROVED** [review]

**Comment by weeman1337** [comment]
> I've finally had some time to mull this over in my head and I think your approach is viable together with a small change in the way the results are displayed that I've committed to your branch just now. Let me know what you think :)

Thanks. Go from my side ✓

File: src/components/BookmarksList.vue [hunk 1]
```diff
@@ -7,19 +7,42 @@
 		<CreateBookmark v-if="newBookmark" />
 		<CreateFolder v-if="newFolder" />
 		<template v-if="$route.name === routes.FOLDER || $route.name === routes.HOME">
-			<Folder
-				v-for="folder in folderChildren"
-				:key="'f' + folder.id"
-				:folder="folder" />
+			<!-- FOLDER VIEW WITH CUSTOM SORTING -->
+			<template v-if="$store.state.settings.sorting === 'index'">
```
Comment (weeman1337) [inline src/components/BookmarksList.vue:12]: Wouldn't it be better to sort somewhere in the "frontend backend" ;)  
Like by using a getter of the store.

File: src/components/BookmarksList.vue [hunk 2]
```diff
@@ -7,19 +7,42 @@
...
+				<template v-for="item in children">
+					<Folder
+						v-if="item.type === 'folder'"
+						:key="item.type + item.id"
+						:folder="$store.getters.getFolder(item.id)[0]" />
```
Comment (weeman1337) [inline src/components/BookmarksList.vue:17]: In my opinion the store shouldn't be accessed from the template.

**Pull request merged**